### PR TITLE
Batch deployment integration with AACS for stable diffusion models

### PR DESCRIPTION
**Pull request by gauravrajguru**
# Description
First two commits are the direct copy of LLama notebook and batch deployment wrapper from batch team.

In this notebook, we demonstrated passing of credential just by using environment variable unlike online deployment (where we demoed both approach using env variable and user assigned identity). This is inline with LLama model batch deployment notebook.  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
- [x] I have coordinated with the docs team (mldocs@microsoft.com) if this PR deletes files or changes any file names or file extensions.
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The public registry name contains text-to-image models\n",
+    "registry_name = \"azureml-staging\"\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:42]:
```suggestion
    "registry_name = \"azureml\"\n",
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+    "\n",
+    "# Name of the text-to-image model to be deployed\n",
+    "model_name = \"runwayml-stable-diffusion-v1-5\"\n",
+    "\n",
+    "import time\n",
+    "\n",
+    "timestamp = str(time.time())\n",
+    "endpoint_name = f\"text-to-image-ep-{timestamp}\"[:32]  # Replace with your endpoint name,\n",
+    "deployment_name = \"default\"  # Replace with your deployment name, lower case only!!!\n",
+    "sku_name = \"STANDARD_NC6S_V3\"  # Name of the sku(instance type) Check the model-list(can be found in the parent folder(inference)) to get the most optimal sku for your model (Default: Standard_DS2_v2)\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:52]:
```suggestion
    "sku_name = \"STANDARD_NC6S_V3\"  # Name of the sku(instance type) Check the model-list(can be found in the parent folder(inference)) to get the most optimal sku for your model (Default: STANDARD_NC6S_V3)\n",
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+    "\n",
+    "print(f\"Connected to workspace {workspace}\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 1.7 Assign variables for Azure Content Safety\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:179]:
```suggestion
    "#### 1.7 Assign variables for Azure AI Content Safety\n",
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+    "Currently, Azure AI Content Safety is in a limited set of regions:\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:180]:
```suggestion
    "Currently, Azure AI Content Safety is supported in a limited set of regions:\n",
```

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        ".pqt": load_parquet,
+    }
+
+    endpoint = os.environ.get("CONTENT_SAFETY_ENDPOINT")
+    key = os.environ.get("CONTENT_SAFETY_KEY")
+    # Create an Content Safety client
+    aacs_client = ContentSafetyClient(endpoint, AzureKeyCredential(key))
+
+
+def get_input_schema(model_path):
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:69]: doc string

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+            "No signature present in MLmodel file: "
+            + ml_model.__str__()
+            + "\nSignatures are encouraged, but AzureML will take a "
+            + "best effort approach to loading the data files."
+        )
+    except Exception as e:
+        raise Exception("Error reading model signature: " + ml_model.__str__() + str(e))
+
+
+def get_dtypes(schema):
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:111]: doc string

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+            return column_dtypes
+    except Exception as e:
+        raise Exception(
+            "Error reading types from model signature schema: "
+            + schema.__str__()
+            + str(e)
+        )
+
+
+def _get_data_loader(file_extension):
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:134]: doc string

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+        return np.genfromtxt(data_file, dtype=g_dtypes)
+    else:
+        g_logger.info(
+            "Reading csv input file into pandas dataframe because the model specifies column based input signature or no signature."
+        )
+        return pd.read_csv(data_file, engine="c", dtype=g_dtypes)
+
+
+def _load_image_as_array(data_file):
+    """
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:165]: Let's ensure to use same docstring style across the board. We have two styles within this file.

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+
+    Returns:
+        np.ndarray: The image loaded as an array.
+    """
+    data = Image.open(data_file)
+    return np.array(data)
+
+
+def _load_image_as_bytes(data_file):
+    """
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:179]: doc string style

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+        data_file (string): The mounted path of the image file.
+
+    Returns:
+        bytes: The contents of the file.
+    """
+    with open(data_file, "rb") as f:
+        return f.read()
+
+
+def load_image(data_file):
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:192]: missing doc string

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+                "more than one input column or a column that is not of dtype "
+                "'binary' is not yet supported.\n"
+                f"This MLflow signature has {len(data_type)} inputs of types: {data_type}."
+            )
+
+    loading_message(data_shape, data_type)
+    return data_array.astype(data_type).reshape(data_shape)
+
+
+def load_parquet(data_file):
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:241]: missing docstring

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+                    result[idx] = _get_jsonable_obj(
+                        model_output, pandas_orient="records"
+                    )
+            return result
+
+    return wrapper
+
+
+@input_filelist_decorator
+def run(batch_input):
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:371]: missing doctring

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py [hunk 13]
```diff
@@ -0,0 +1,584 @@
+# Copyright (c) Microsoft. All rights reserved.
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:1]: For the missing doc strings - I am assuming, that how it was at the source.

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 14]
```diff
@@ -0,0 +1,693 @@
...
+    ")\n",
+    "score_df.head()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The output file above will contain one line for each file, and each line will have multiple arrays corresponding to each line of the file. If you see [], that means ACS has stripped the response from the model"
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:664]:
```suggestion
    "The output file above will contain one line for each file, and each line will have multiple arrays corresponding to each line of the file. If you see [], that means AACS has stripped the response from the model."
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 15]
```diff
@@ -0,0 +1,693 @@
...
+    "    - scoring_job.creation_context.created_at,\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The outputs generated by the deployment job will be placed in an output named `score`:"
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:627]:
```suggestion
    "The outputs generated by the deployment job will be placed in an output named `score`."
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 4.4 Download the results\n",
+    "\n",
+    "The deployment creates a child job that executes the scoring. We can get the details of it using the following code:"
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:595]:
```suggestion
    "The deployment creates a child job that executes the scoring. We can get the details of it using the following code."
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 17]
```diff
@@ -0,0 +1,693 @@
...
+   "source": [
+    "ml_client.jobs.get(job.name)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can wait for the job to finish using the following code:"
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:576]:
```suggestion
    "We can wait for the job to finish using the following code."
```

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 18]
```diff
@@ -0,0 +1,693 @@
...
+    "job = ml_client.batch_endpoints.invoke(endpoint_name=endpoint.name, input=input)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 4.3 Get the details of the invoked job\n",
+    "Let us get details and logs of the invoked job"
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:559]:
```suggestion
    "Let us get details and logs of the invoked job."
```

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py (hunk 13 above)
Comment (gauravrajguru) [reply sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:1]: Regarding missing doc string and different format, this is coming from the source (i..e, original file from batch team wrapper.) Can we keep it as it is for original code. I have added docstring wherever I have touched in the code.

File: sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py (hunk 8 above)
Comment (gauravrajguru) [reply sdk/python/foundation-models/system/inference/text-to-image/scoring-files/score/score_batch.py:165]: This docstring was already present in original code of batch deployment team.

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 19]
```diff
@@ -0,0 +1,693 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# How to create an Azure AI Content Safety (AACS) enabled text-to-image batch endpoint (Preview)\n",
```
Comment (vivek-dani) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:8]: What do you think about framing it as "Create an Azure AI Content Safety (AACS) enabled text-to-image batch endpoint (Preview) .." ?
similar to other notebooks we have

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 20]
```diff
@@ -0,0 +1,693 @@
...
+    "### This notebook will walk you through the steps to create an __Azure AI Content Safety__ enabled __text-to-image__ batch endpoint.\n",
+    "### This notebook is under preview\n",
+    "### The steps are:\n",
+    "1. Create an __Azure AI Content Safety__ resource for moderating the request from user and response from the __text-to-image__ batch endpoint.\n",
+    "2. Create a new __Azure AI Content Safety__ enabled __text-to-image__ batch endpoint with a custom score.py which will integrate with the __Azure AI Content Safety__ resource to moderate the response from the __text-to-image__ model and the request from the user. To make the custom score.py to successfully authenticated to the __Azure AI Content Safety__ resource, use __Environment variable__ to pass the access key of the __Azure AI Content Safety__ resource to the custom score.py via environment variable. The custom score.py can use the key directly to access the Azure AI Content Safety resource, this option is less secure, if someone in your org has access to the endpoint, he/she can get the access key from the environment variable and use it to access the Azure AI Content Safety resource.\n",
```
Comment (vivek-dani) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:13]: 1) Can we break this in to 2 sentences as below:

> Create a new Azure AI Content Safety enabled text-to-image batch endpoint with a custom score.py script. This script will integrate with the AACS resource to moderate the response from the text-to-image model and the request from the user. 

2) We can define acronym for Azure AI Content Safety (AACS) once and then use acronym every else.

Reducing sentence size will make it more understandable. This just my perception; happy to discuss if needed.

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 21]
```diff
@@ -0,0 +1,693 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1. Prerequisites\n",
+    "#### 1.1 Check List:\n",
+    "- [x] The identity you are using to execute this notebook(yourself or your VM) need to have the __Contributor__ role on the resource group where the AML Workspace your specified is located, because this notebook will create an Azure AI Content Safety resource using that identity.\n"
```
Comment (vivek-dani) [inline sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:24]: we can remove the bullet point, as there is only 1 point

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb (hunk 20 above)
Comment (gauravrajguru) [reply sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:13]: Thanks. Addressed it,

[the remark this answers is not in the document]
Comment (gauravrajguru) [reply sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb]: Addressed it.

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 22]
```diff
@@ -0,0 +1,693 @@
...
+    "#### 1.3 Install Dependencies(as needed)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# uncomment the following lines to install the required packages\n",
```
[the remark this answers is not in the document]
Comment (gauravrajguru) [reply sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb]: Changed to have uncommented as default.

File: sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb [hunk 23]
```diff
@@ -0,0 +1,693 @@
...
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 3.4 Create text-to-image batch endpoint\n",
+    "This step may take a few minutes."
+   ]
+  },
```
[the remark this answers is not in the document]
Comment (gauravrajguru) [reply sdk/python/foundation-models/system/inference/text-to-image/safe-text-to-image-batch-deployment.ipynb:397]: made it similar to these notebooks are close as possible.

**Review by vivek-dani — APPROVED** [review]

**Pull request merged**